### PR TITLE
refactor(compiler-cli): add async-await transform

### PR DESCRIPTION
**Pull request by petebacondarwin**
See individual commits.

### CODE REVIEW

**Review by devversion — APPROVED** [review]
LGTM on the dev-infra changes. I couldn't get a lint error with the snippet in b5c6b799810724f110206708669c5dbf07494f44 but I ran into issues with semicolons for such bound methods as well.

File: packages/compiler-cli/test/compliance/test_cases/async_transform/typescript_baselines/asyncWithVarShadowing_es6.js [hunk 1]
```diff
@@ -0,0 +1,272 @@
+function fn1(x) {
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_cases/async_transform/typescript_baselines/asyncWithVarShadowing_es6.js:1]: I don't see why the generator hasn't been hoisted in these cases?

**Review by JoostK — COMMENTED** [review]
This looks really well put together 💯 The comments were extremely helpful to follow what is going on.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  /**
+   * Implement this method to create a new instance of Scope based on the given `node`.
+   */
+  abstract createScope(_node: ts.Node): Scope|null;
+
+  /**
+   * Push a scope onto the stack.
+   *
+   * `null` is used to capture a lexical scope that overrides the previous one but does not contain
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts:34]: I'll move the comment to `createScope()`.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * A stack of lexical scopes.
+ *
+ * Each time a visitor enters a node that creates a new lexical scope then a new scope for that node
+ * should be pushed into the stack.
+ *
+ * When the visitor leaves a node that has a lexical scope then the scope should be popped off the
+ * stack.
+ */
+export abstract class BaseScopes<Scope> {
+  private uniqueIds: Record<string, number> = {};
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts:21]: My understanding is that `Map` has a small constant overhead above a simple object and that there may be lots of these scope objects that need such a `uniqueIds` mapping.

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts:21]: I should though change it to have no prototype:

```suggestion
  private uniqueIds: Record<string, number> = Object.create(null);
```

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts:21]: You are certainly right that we don't need `BaseScope`.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+   */
+  abstract createScope(_node: ts.Node): Scope|null;
+
+  /**
+   * Push a scope onto the stack.
+   *
+   * `null` is used to capture a lexical scope that overrides the previous one but does not contain
+   * any relevant information.
+   */
+  push(node: ts.Node): void {
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts:37]: I think I did try that but it made things more complicated. Let me try again.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    this.scopeStack.push(this.createScope(node));
+  }
+
+  /**
+   * Returns the scope at the top of the stack.
+   */
+  peek(): Scope|null {
+    return this.scopeStack[this.scopeStack.length - 1] ?? null;
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts:45]: OK, the code isn't as pretty then :-)

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+/**
+ * Test to see if the given `node` is an `arguments` identifier within an "async arrow" function.
+ */
+export function isArgumentsAccessInAsyncArrowFn(node: ts.Node): node is ts.Identifier {
+  if (!ts.isIdentifier(node) || node.text !== 'arguments') {
+    return false;
+  }
+  let current = ts.getOriginalNode(node);
+  while (current && !ts.isBlock(current) && !ts.isArrowFunction(current)) {
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts:114]: The `ArgumentsScope` is created by a "non-arrow" function, which may or may not be async (it doesn't matter).

What we are checking here is whether the use of `arguments` was inside an async arrow function, which is within the `ArgumentsScope`.

For example:

function foo() {
  const a => () => /*...*/ arguments /*...*/ ;
  const b => () => {
    /*...*/
    arguments
    /*...*/
  };
  const async c => () => /*...*/ arguments /*...*/ ;
  const async d => () => {
    /*...*/
    arguments
    /*...*/
  };
}
```

In this case the `ArgumentsScope` is created when we enter the body of `foo()`.
Then when we hit each of the `arguments` keywords, we need to see whether we are inside an async arrow function.
In this case, it is true for `c` and `d` but not `a` or `b`.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+        argumentsScope.hasArgumentsAccess) {
+      // This `arguments` container contains an `arguments` access from within an async arrow
+      // function.
+      // Define the `ɵarguments` proxy variable at the start of the function body.
+      node = this.addArgumentsProxy(node, argumentsScope);
+    }
+
+    if (ts.isAwaitExpression(node)) {
+      // Convert `await` expression to `yield` expression.
+      node = this.factory.createYieldExpression(undefined, node.expression);
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:140]: There are 39 of these `undefined` arguments in this file.
Personally, in the case of creating these AST object, I find adding comments for each makes the code less readable.

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:140]: I'll compromise by creating named constants that alias `undefined`.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+   * to the `Zone.__awaiter()` helper.
+   *
+   * The async function is replaced with an updated function, which runs a generator function
+   * inside a call to the `Zone.__awaiter()` helper function.
+   *
+   * This function uses a specific TypeScript factory function for each type of function that can be
+   * transformed.
+   */
+  private transformAsyncFunction(
+      asyncFn: ts.SignatureDeclaration,
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:221]: Actually that highlighted an issue with `AsyncFunction` which was missing `ArrowFunction` from its union.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+      asyncSuperScope: AsyncSuperScope|null): ts.SignatureDeclaration {
+    const modifiers = filterAsyncModifier(asyncFn.modifiers);
+    const generatorFn = this.createGeneratorFunction(asyncFn, asyncSuperScope);
+    const args = this.convertParamsToArgs(asyncFn.parameters ?? []);
+    const awaiterCall = this.createAwaiterCall(generatorFn, args);
+
+    if (ts.isArrowFunction(asyncFn)) {
+      // No proxy helpers needed, so just update the arrow function
+      return this.factory.updateArrowFunction(
+          asyncFn, undefined, asyncFn.typeParameters, asyncFn.parameters, asyncFn.type,
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:231]: Yes you are right. I missed this one. It should be `modifiers`.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+  private transformAsyncFunction(
+      asyncFn: ts.SignatureDeclaration,
+      asyncSuperScope: AsyncSuperScope|null): ts.SignatureDeclaration {
+    const modifiers = filterAsyncModifier(asyncFn.modifiers);
+    const generatorFn = this.createGeneratorFunction(asyncFn, asyncSuperScope);
+    const args = this.convertParamsToArgs(asyncFn.parameters ?? []);
+    const awaiterCall = this.createAwaiterCall(generatorFn, args);
+
+    if (ts.isArrowFunction(asyncFn)) {
+      // No proxy helpers needed, so just update the arrow function
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:229]: OMG! I can't even remember why!
I'll dig into this and make sure there is a good comment.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 11]
```diff
@@ -0,0 +1,583 @@
...
+          asyncFn.parameters, asyncFn.type, body);
+    }
+
+    if (ts.isMethodDeclaration(asyncFn)) {
+      return this.factory.updateMethodDeclaration(
+          asyncFn, asyncFn.decorators, modifiers, undefined, asyncFn.name, asyncFn.questionToken,
+          asyncFn.typeParameters, asyncFn.parameters, asyncFn.type, body);
+    }
+
+    throw new Error('Unsupported async function type');
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:267]: For some reason it doesn't. It just types `asyncFn` as `never`. And if I remove the `throw` statement then it complains about a missing return!

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+        text: `
+          const ${asyncScope.elementAccessProxy.text} = (function (geti, seti) {
+            const cache = Object.create(null);
+            return name => cache[name] || (cache[name] = { get value() { return geti(name); }, set value(v) { seti(name, v); } });
+          })(name => super[name], (name, value) => super[name] = value);`,
+      });
+    } else if (asyncScope.hasElementRead) {
+      ts.addEmitHelper(body, {
+        name: 'super-element-read-proxy',
+        scoped: true,
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:290]: That is a mistake. It only needs to be `scoped: true` if we want to use `((node: EmitHelperUniqueNameCallback) => string)` for the `text` property. But since we only use a plain string, then we can have both as `scoped: false`.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 13]
```diff
@@ -0,0 +1,583 @@
...
+        });
+        return this.factory.createArrayLiteralExpression(args);
+      }
+      if (ts.isObjectBindingPattern(node)) {
+        const args = node.elements.map(element => {
+          const name = visitBinding(element.name);
+          if (element.dotDotDotToken !== undefined) {
+            return this.factory.createSpreadAssignment(name);
+          }
+          return (element.propertyName === undefined && ts.isIdentifier(name)) ?
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:460]: Yeah you are right. It is feasible (I think) for `propertyName` to be undefined while `name` is not an identifier.
I'll see if I can work out a test that breaks this

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:460]: Right now I can't actually think of an example where the property name is missing and the `name` is not an identifier.
So perhaps a more accurate type assertion would be:

```ts
          return (element.propertyName === undefined) ?
              this.factory.createShorthandPropertyAssignment(name as ts.Identifier) :
              this.factory.createPropertyAssignment(element.propertyName, name);
```

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 14]
```diff
@@ -0,0 +1,583 @@
...
+
+  /**
+   * Compute a unique name for the new generator function.
+   *
+   * The name is computed from the name of the `async` function if this is available.
+   */
+  private computeGeneratorName(asyncFn: ts.SignatureDeclaration): ts.Identifier {
+    // We need to use the original node because if this asyncFn has been synthesized it might not
+    // have a `parent` property.
+    const node = ts.getOriginalNode(asyncFn) as ts.SignatureDeclaration;
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:532]: I didn't know about that overload.
It turns out there is no `ts.isSignatureDeclaration` but we can actually use `isAsyncFunction` and the overload guarantees that it is not `undefined` too!

File: packages/compiler-cli/src/ngtsc/async_transform/src/super_accesses_transform.ts [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+   *
+   * - property read access: `bar = super.foo` -> `bar = ɵsuper.foo`
+   * - property write access: `super.foo = bar` -> `ɵsuper.foo = bar`
+   * - element read access: `bar = super['foo']` -> `ɵsuperIndex('foo')` or
+   *   `ɵsuperIndex('foo').value`
+   * - element write access: `super['foo'] = bar` -> `ɵsuperIndex('foo').value = bar`
+   * - method calling: `super.foo()` -> `ɵsuper.foo.call(this)`
+   * - element calling: `super[exp]()` -> `ɵsuperIndex(exp).call(this)`
+   */
+  private transformSuperAccesses(node: ts.Node, superScope: AsyncSuperScope): ts.Node {
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/super_accesses_transform.ts:58]: Indeed! And also future transforms will need to be accounted for if we allow this transform to be used on code that targets newer syntaxes...

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts]: Seems to work fine. Thanks for bringing it up.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts]: @JoostK what do you think of this?

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts:116]: Does this make sense @JoostK or did I misunderstand your comment?

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts]: One thing I noticed, I think, was that if you set the compiler config not to emit helpers then I think it will also fail to emit these two helpers. If this is true then this might not be a viable solution, since these helpers cannot be imported from elsewhere...

File: packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts [hunk 16]
```diff
@@ -388,4 +388,15 @@ export interface MiscOptions {
    * Disable TypeScript Version Check.
    */
   disableTypeScriptVersionCheck?: boolean;
+
+  /**
+   * Transform async-await functions into Zone.js compatible generator functions.
+   *
+   * Native async-await functions cannot be hooked by Zone.js, which can lead to missing change
+   * detection notifications.
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:396]: Done.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts:116]: I was exploring the idea of having `ArgumentsScope.isAsync` to tell whether it's an async block or not, instead of traversing the AST upwards to determine that. But perf is probably non-critical in this case since the traversal only occurs for occurrences of `arguments`.

On a tangent, I just realized that we're not testing if the `arguments`/`super` identifiers are bare reads, or if they are used as property accesses (`foo.arguments` als has a `ts.Identifier` for `arguments` that we shouldn't touch). Or am I missing something this early in the morning? :-)

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/base_scope.ts]: I still like having something like `private uniqueSuper = new UniqueId('super');` and then referencing that field instead of the more dynamic approach with a `Map`/`{}`.

I'm not aware of any overhead of Maps vs objects in this particular case; `Map` is always monomorphic vs an object that has to deal with insertion order of properties, but then an actual lookup in the Map is not monomorpic versus a polymophic/megamorphic read when using an object. I'm not sure about memory usage, an object is probably cheaper? Anyway, your approach does have the benefit of lazily allocating the unique counters, which my suggestion with `UniqueId` does not.

I guess bottom line is that any perf/size overhead might be meaningless without benchmarks, which I don't think is worth it at this point. So I'm okay with either approach, the `Object.create(null)` approach is also fine with me.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:238]: I think I just copied this from the TS implementation, since I have no recollection of why it is this way.
Furthermore, I tried out the TS implementation and it actually generates invalid JS - see [playground](https://www.typescriptlang.org/play?noImplicitAny=false&target=3#code/MYGwhgzhAEBCkFNoG8BQ1qQJ4DtjQDMB7IgCjABpoAjAShXQ2mCJwiJAQDoQiBzUgHJ4EBINoBuRgF9Us0JBgBhABYBLEABNoCAB4AXBDk0wRSNBl3kqdBk2asI+wiWgBeTBFz5S9NwD5MAHcwNWcIAFcABwQAJy5iMkoaSUYMRN8pDFkMbDwXJJt6CyYWNg5uXgFBVQ1NcSzoWXlHZ3wPHAQg6FqtTNRgLisARgAGKgAmUckgA)

```ts
class Child extends Base {
  x(a, b) {
    const foo = async () => await super.foo(a, b);
    foo();
  }
}
```

generates

```ts
    x(a, b) {
        const foo = () => __awaiter(this, void 0, void 0, function* () { return yield super.foo(a, b); });
        foo();
    }
```

And when to try to run this the JS engine complains that super is not valid in this place

```
[ERR]: "Executed JavaScript Failed:" 
[ERR]: 'super' keyword unexpected here 
```

If I run the same code targeting ES2018, then it works fine. So the original TS is valid, but not the downleveled code.
So I think that we should indeed be creating the `super` proxy despite what TS is doing.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts:116]: Yes, and moreover, we would have to check for arrows and non-arrows on each of the original traversal steps to change the "mode" of the current node as we moved in and out of arrow functions, so I think that we would actually end up doing much more work by tracking this rather than backtracking up the ancestors on the occasions when it is necessary.

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts:116]: > On a tangent, I just realized that we're not testing if the arguments/super identifiers are bare reads, or if they are used as property accesses (foo.arguments also has a ts.Identifier for arguments that we shouldn't touch). Or am I missing something this early in the morning? :-)

Yes you are (once again) correct. We must also check for `PropertyAssignment` as well as `PropertyAccessExpression` parents. E.g.

```ts
foo.arguments = 200;
foo = { arguments: 300 };
```

Furthermore we have to deal with the awkward `ShorthandPropertyAssignment`:

```ts
foo = { arguments }
```

which needs to be transformed into a normal property access:

```ts
foo = { arguments: arguments }
```

And then have the standard processing...

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:276]: I can get it to work if I just use an `else` for the last item.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts]: This is actually also true of the TS transformed helpers...

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:238]: It turned out that this was a bit more complicated since we need to update `super` containers with proxies even if they are not async but only if they contain `super` accesses within an async function inside the `super` container. 😱

File: packages/compiler-cli/test/compliance/test_cases/async_transform/typescript_baselines/asyncWithVarShadowing_es6.js (hunk 1 above)
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/test_cases/async_transform/typescript_baselines/asyncWithVarShadowing_es6.js:1]: LOL! This was not listed in the `TEST_CASES.json` file so it was not being tested.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts]: Added processing for property access and assignment false positives.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 17]
```diff
@@ -0,0 +1,583 @@
...
+   * Normally, the generator function will be hoisted to the containing lexical scope and the name
+   * of the function will be returned so that it can be referenced in the `Zone.__awaiter()` call.
+   *
+   * If the lexical scope contains access to `super` that need to be transformed, then instead of
+   * hoisting the generator function is returned so that it can be injected inline into the
+   * `Zone.__awaiter()` call.
+   */
+  private createGeneratorFunction(
+      asyncFn: ts.SignatureDeclaration, asyncScope: AsyncSuperScope|null): ts.Expression {
+    if (asyncScope === null || !asyncScope.hasSuperAccess) {
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:544]: This is true, but equally I don't think there is a downside to hoisting these - so for now I am going to punt on adding that additional logic. We could add it as a follow-up if it is needed.

File: packages/compiler-cli/src/ngtsc/async_transform/src/super_accesses_transform.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/super_accesses_transform.ts:60]: Optional chaining was mostly supported for free. I just needed to tweak the conversion to a `.call()` expression for super calls, since we need to move the `.?` over from the call to the property access.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts [hunk 18]
```diff
@@ -107,27 +109,52 @@ export function isArgumentsContainer(node: ts.Node): node is ArgumentsContainer
...
+/**
+ * Test to see if this node is directly within an arrow function.
+ */
+function isInAsyncArrowFn(node: ts.Node): boolean {
+  while (node.parent !== undefined) {
+    if (ts.isArrowFunction(node)) {
+      // We have an arrow function with no block: `(arg1, arg2) => expression`
+      return isAsync(node);
+    }
+    if (ts.isBlock(node) && ts.isArrowFunction(node.parent)) {
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts:153]: Is this logic really needed? If we unroll the loop one more iteration we'll naturally hit the `ts.isArrowFunction` case above.

File: packages/compiler-cli/src/ngtsc/async_transform/src/utils.ts [hunk 19]
```diff
@@ -99,3 +99,130 @@ export function isNamedFunction(node: ts.SignatureDeclaration): node is ts.Signa
 
   return false;
 }
+
+/**
+ * Test to see if the `node` is the name part of a declaration.
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/async_transform/src/utils.ts:104]: TBH I haven't checked the extensiveness of all of these. The logic LGTM and any missing cases here are very unlikely to be hit.

Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/async_transform/src/utils.ts:104]: I was wondering about `const { arguments } = { arguments: {} };` but TypeScript doesn't even allow me to type that in strict mode, reporting TS1210 on the destructuring assignment.

**Review by JoostK — COMMENTED** [review]
I noticed there is a bug with hoisting the generator when an async function is within a block:

```ts
{
  let a = 1;
  (async () => {
    a++;
  })();
}
```

this generates 

```js
function* anonymous_generator() {
    a++;
}
{
    let a = 1;
    (() => Zone.__awaiter(this, [], anonymous_generator))();
}
```

i.e. the generator is hoisted beyond the scope that declares `a`, so that's now an invalid write.

---

Additionally, some fixup commits seem to be targeting the wrong commit. The last two `fixup! refactor(compiler-cli): expose the new async_transform via a compiler option.` have to fixup `refactor(compiler-cli): implement async-await transform for Zone.js`. Hopefully reordering them below `refactor(compiler-cli): expose the new async_transform via a compiler option.` does not introduce conflicts.

File: packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts (hunk 18 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/scopes/arguments_scope.ts:153]: You are right!

But then thinking about this more I realised that we have to stop at the current "argument container" otherwise we might have a scenario like this:

```ts
const topLevelArrow = async () => {
  function innerNonArrow() {
    const x = arguments[0];
  }
};
```

In this case when we hit `arguments` we would then start traversing up the parents until we hit the async arrow function (`topLevelArrow()`) rather than stopping at the argument container boundary of `innerNonArrow()`.

I'll add some more unit tests and fix this.

**Comment by petebacondarwin** [comment]
Thanks for the review @JoostK - regarding the hoisting bug - good spot!!:

I don't know how to fix that since we don't actually get any control over the TS `hoistFunctionDeclaration()` helper function, which assumes that it can jump all the way up to the nearest function body, rather than stopping at the nearest block.

I looked into changing the generator into a function expression with a variable declaration, but then the declaration gets hoisted without being initialized and we would have to jump through hoops to position the initialization accurately ourselves. In which case we may as well do the same with a function declaration.

Given we are already inlining some scenarios, I propose that we just go with inline definitions all the time. This makes the code simpler and more maintainable, it is also the same as how TS and Babel do it, also it may not have optimal performance, but it will be no worse that what they have now, where these are all downleveled to promises in ES2015.

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts [hunk 20]
```diff
@@ -639,3 +639,50 @@ export class AsyncFunctionVisitor {
...
   }
 }
+
+/**
+ * Can the generator for the given `asyncFn`, within the given `asyncScope`, be hoisted to the top
+ * of the current lexical scope?
+ */
+function canBeHoisted(asyncFn: AsyncFunction, asyncScope: AsyncSuperScope|null): boolean {
+  asyncFn = ts.getOriginalNode(asyncFn, isAsyncFunction);
+  return !asyncScope?.hasSuperAccess &&
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:649]: The `?.` should not be needed here

**Review by JoostK — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts (hunk 20 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:649]: The asyncScope can be null

Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/async_transform/src/async_transform.ts:649]: D'oh! I read it as `asyncFn` which can't be nullable.
